### DISCUSSION TITLE
Comment about Linux from session discussion

### DISCUSSION THREAD

**Discussion by laat0003** [gmdsi/GMDSI_notebooks#97] [General]
Hi,

I hope you don't mind me asking this here seeing as it is not strictly to do with the notebooks but was mentioned in today's session. The comment about @mnfienen convincing everyone to use Linux piqued my interest. I do most of my work with Windows but do run Ubuntu VM's regularly for other purposes.  I understand the benefits of Linux from an everyday computer user point of view but I'm curious to know if there are specific benefits for a groundwater modeler, keeping in mind I work in consulting where every man and his dog uses Windows.

Cheers,
Tariq

**Comment by mnfienen** [comment]
Hey @laat0003  - I'm not sure what the comment was since I wasn't in the Australia/NZ session, but I recognize the challenge of moving away from Windows in consulting. We have similar pressures in government where I work as well. The main advantage I often point out, though, is in cases where massive parallel runs are required. From networking and licensing standpoints, it can be easier to work on Linux. Particularly if you run in the cloud - the cost of Windows VMs in cloud is more expensive.

I sometimes may be a little aggressive in my recommendations - Windows is perfectly fine for most work as well. But when it comes to big computing jobs, there are some advantages.

Cheers
Mike

**Comment by laat0003** [comment]
Thanks for that @mnfienen. Appreciate your thoughts and the information. I haven't worked anywhere that uses cloud computing yet so wasn't aware that Windows VMs incur increased costs. 

Cheers,
Tariq

**Comment by briochh** [comment]
Apologies for using your name in vain @mnfienen! This was something of a flippant remark and I was definitely projecting! @laat0003, as you point out there are challenges with moving away from windows when our clients might only be comfortable in that environment. 

However, as @mnfienen points out *nix might be a nicer option where we a looking at heaps of parallel runs (e.g. HTCondor set up is much easier than on Windows). I have also recently noticed significant performance variations between *nix based systems and Windows for both PEST++ and MODFLOW (related to IO for super wide tabular files, although there are workaround.).

I think it has become easier to switch between OS systems, notwithstanding the obvious caveat that model results might vary slightly! Generally I have found that I can run workflows developed in python on both my Mac and on Windows, without too much hassle.

**Comment by laat0003** [comment]
Thanks @briochh. I'm curious about what you both consider to be "lots" of parallel runs.  This may differ for many people depending on previous modelling experience with PEST/PEST++. Moreover, do you have a cap on the amount of time that you are willing to devote to a UA? If so, does this affect your model design and/or parameterization approach? These questions are perhaps a bit off the Linux topic but I am trying to build some intuition around what other modelers are doing just so that I know were on the same page. That said I'm sure there are others who may be interested in the answers to these questions so it could be a useful discussion point in a session.

Cheers,
Tariq

**Comment by mnfienen** [comment]
Hey @laat0003  - this is great discussion (although agree we are straying from the *nix/windows discussion ;) ).

But fwiw I tent to start ensembles with 500 realizations (or more) and being impatient, I also tend to want one worker per realization to be sure I get results in approximately one run time per ensemble evaluation or lambda search. So....yeah - hundreds is the norm I'd say. And, we've run on 300 - 500 Windows machines in the past, but the *nix path involves fewer tears in the end.

HTH
Mike

**Comment by laat0003** [comment]
Thanks @mnfienen, 

This is a real eye opener for me so greatly appreciate the comments.

Wow! That is a lot of agents in comparison to what I've used previously. I think the most I've ever used at once is 40 and also usually target 500 to 700 realizations for an ensemble (haven't used fosm in a while). I'm usually doing whatever I can with the model design to make sure it gets through 3 to 4 smoothing iterations in a few hours. 

Cheers,
Tariq

**Comment by robinkeegan** [comment]
Thanks @mnfienen and @laat0003, 

That is very insightful! In consultancy, I have done most of my modelling on a desktop that is limited to about 13 parallel agents, which often means I am waiting days. I can definitely see the benefits in running all realizations at once in terms of less time waiting. Has there been an effort to create a tutorial/online resource describing the setup of a cloud computing environment for pestpp/modflow related modelling? I have looked into it recently and found it is quite involved, which would be a barrier for most people.  In the end I managed to setup an azure environment using linux although it was a bit clunky and took a while. I am curious if there is a good online resource already and if not if there would be interest in developing one? I would happy to contribute.

Robin
